### PR TITLE
Introduce switch type in forms

### PR DESCRIPTION
**Pull request by Quetzacoalt91**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Add the switch representation on our new forms (ON / OFF)
| Type?         | new feature
| Category?     | BO
| BC breaks?    | Nope
| Deprecations? | Nope
| Fixed ticket? | /
| How to test?  | Our migrated pages must use the switch instead of a select for On/Off cases

capture du 2018-01-26 11-51-57

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8705)

### CODE REVIEW

File: src/Adapter/Admin/NotificationsConfiguration.php [hunk 1]
```diff
@@ -48,9 +48,9 @@ public function __construct(Configuration $configuration)
     public function getConfiguration()
     {
         return array(
-            'show_notifs_new_orders' => $this->configuration->get('PS_SHOW_NEW_ORDERS'),
-            'show_notifs_new_customers' => $this->configuration->get('PS_SHOW_NEW_CUSTOMERS'),
-            'show_notifs_new_messages' => $this->configuration->get('PS_SHOW_NEW_MESSAGES'),
+            'show_notifs_new_orders' => (bool)$this->configuration->get('PS_SHOW_NEW_ORDERS'),
```
Comment (mickaelandrieu) [inline src/Adapter/Admin/NotificationsConfiguration.php]: we shouldn't have to cast values, if its not boolean values lets change them in database.

File: src/Adapter/Smarty/SmartyCacheConfiguration.php [hunk 2]
```diff
@@ -70,8 +70,8 @@ public function updateConfiguration(array $configuration)
     {
         if ($this->validateConfiguration($configuration)) {
             $this->configuration->set('PS_SMARTY_FORCE_COMPILE', $configuration['template_compilation']);
-            $this->configuration->set('PS_SMARTY_CACHE', $configuration['cache']);
-            $this->configuration->set('PS_SMARTY_LOCAL', $configuration['multi_front_optimization']);
+            $this->configuration->set('PS_SMARTY_CACHE', (bool)$configuration['cache']);
+            $this->configuration->set('PS_SMARTY_LOCAL', (bool)$configuration['multi_front_optimization']);
```
Comment (mickaelandrieu) [inline src/Adapter/Smarty/SmartyCacheConfiguration.php]: same here, I'm not easy with new PHP code fixing possible database issues! How about a $configuration->getBoolean('KEY') instead?

File: src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/bootstrap_3_layout.html.twig [hunk 3]
```diff
@@ -256,6 +256,19 @@
     {% endspaceless %}
 {% endblock date_picker_widget %}
 
+{% block switch_widget %}
+    {% spaceless %}
+    <span class="ps-switch">
+        {% for choice in choices %}
+            {% set name = id ~'_' ~ choice.value %}
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/bootstrap_3_layout.html.twig]:
```diff
- name
+ input_id
````

File: src/Adapter/Smarty/SmartyCacheConfiguration.php (hunk 2 above)
Comment (Quetzacoalt91) [reply src/Adapter/Smarty/SmartyCacheConfiguration.php]: Ah, realy nice suggestion here!

Comment (Quetzacoalt91) [reply src/Adapter/Smarty/SmartyCacheConfiguration.php]: We could reuse the same features as the ParameterBag from Symfony, but as an interface with the database. https://api.symfony.com/3.0/Symfony/Component/HttpFoundation/ParameterBag.html

File: src/Adapter/Admin/NotificationsConfiguration.php (hunk 1 above)
Comment (Quetzacoalt91) [reply src/Adapter/Admin/NotificationsConfiguration.php]: I had to, as the Configuration table allows every primitive types to be stored.

File: src/Adapter/Smarty/SmartyCacheConfiguration.php (hunk 2 above)
Comment (mickaelandrieu) [reply src/Adapter/Smarty/SmartyCacheConfiguration.php]: agree :)

**Comment by Quetzacoalt91** [comment]
@eternoendless @mickaelandrieu can you have a look at the commit https://github.com/PrestaShop/PrestaShop/pull/8705/commits/2d96cd5d7f62468dd483b9b1d0cb98788723c39c?

Thanks

**Comment by eternoendless** [comment]
Reviewed 14 of 15 files at r1, 1 of 1 files at r2, 10 of 13 files at r3, 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/Adapter/Configuration.php, line 43 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8705#-L5mrr3yUciKL8H2vyTY:-L5mrr3yUciKL8H2vyTZ:b-s794vs) ([raw file](https://github.com/prestashop/prestashop/blob/a086da03cc64e2ceac007c55ef66c5ee3a4b191f/src/Adapter/Configuration.php#L43)):*
> ```PHP
>     {
>         // Do nothing
>         if (count($parameters)) {
> ```

I think that it's generally better to use `!empty()` instead of `count()` when you are trying to know if an array is empty. Not only because it makes what you're testing for clearer, but there's also a performance difference (albeit tiny) in favor of `empty` since it's a language construct, whereas `count` instead of a function call.

---

*[src/PrestaShopBundle/Form/Admin/Type/SwitchType.php, line 51 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8705#-L5nBgWpfkJYcG7Td706:-L5nBgWpfkJYcG7Td707:bj31dcz) ([raw file](https://github.com/prestashop/prestashop/blob/a086da03cc64e2ceac007c55ef66c5ee3a4b191f/src/PrestaShopBundle/Form/Admin/Type/SwitchType.php#L51)):*
> ```PHP
>             'choices' => array(
>                 'No' => false,
>                 'Yes' => true,
> ```

Are these values translated automagically?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8705)*

**Comment by Quetzacoalt91** [comment]
*[src/PrestaShopBundle/Form/Admin/Type/SwitchType.php, line 51 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8705#-L5nBgWpfkJYcG7Td706:-L5nXMjUyqYKTOiqHA49:bz0asx1) ([raw file](https://github.com/prestashop/prestashop/blob/a086da03cc64e2ceac007c55ef66c5ee3a4b191f/src/PrestaShopBundle/Form/Admin/Type/SwitchType.php#L51)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Are these values translated automagically?
</blockquote></details>

Yes they are, look at the "choice_translation_domain" key :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8705)*

**Comment by eternoendless** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8705#-:-L5nt0aH5KVVqzFPyNX8:bnfp4nl)*

**Comment by eternoendless** [comment]
Code looks good but travis build is failing

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8705#-:-L5ntY6p585gUdtw87K7:betsee3)*

**Comment by Quetzacoalt91** [comment]
I had to remove all occurrences of  `bootstrap.php.cache` in order to make the tests pass. Really weird.

**Comment by eternoendless** [comment]
Weird indeed, why did we have that in the first place?

---

Reviewed 1 of 1 files at r5, 7 of 7 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8705#-:-L5rxNp9VyvLqM6ELMBY:b8fmvyh)*

**Review by eternoendless — DISMISSED** [review]

**Comment by mickaelandrieu** [comment]
is the file still generated with php 5.6? it contains inline definitions of some classes/services to speed up the boot of Symfony application (not specific to PrestaShop). It's useless in PHP7+, because others optimizations are done in this case

**Comment by marionf** [comment]
@Quetzacoalt91 

I don't see the swith on the maintenance page
capture du 2018-03-13 16-26-23

And some "No" are in grey and others in white. If I click on the grey one, it becomes white.
Can we have all in the same color ?
capture du 2018-03-13 16-27-02

**Comment by Quetzacoalt91** [comment]
That's because the maintenance page PR was merged after this one was created. :)

I will apply the change on this one as well.

**Review by mickaelandrieu — DISMISSED** [review]

**Review by mickaelandrieu — DISMISSED** [review]

**Comment by marionf** [comment]
@Quetzacoalt91 

Some buttons are ok but others are still grey
capture du 2018-03-27 10-23-24

**Comment by Quetzacoalt91** [comment]
Sorry @marionf, I missed the other screenshot. The values were not boolean, that why you had this visual issue

**Review by mickaelandrieu — DISMISSED** [review]

**Comment by marionf** [comment]
@Quetzacoalt91 Can you solve it also on the maintenance page ?

**Comment by Quetzacoalt91** [comment]
Done, thanks :)

**Review by mickaelandrieu — APPROVED** [review]

**Comment by marionf** [comment]
Thank you @Quetzacoalt91 :+1:

**Comment by mickaelandrieu** [comment]
Thanks everyone!

**Pull request merged**